### PR TITLE
Do not remove segments from currentlyMovingSegments in DruidBalancer if move is impossible or not needed

### PR DESCRIPTION
**Pull request by dgolitsyn**
1. Segments that won't be moved by DruidBalancer should not be removed from currentlyMovingSegments explicitly: all moved segments are removed by callbacks or not inserted if  move is not possible or segment move has already been scheduled. 
Currently coordinator stats are corrupted if the same segment was chosen for move within the same tier.
See testMoveSameSegmentTwice for simple example (before this fix movedCount was 0 while 1 segment actually has been moved)

2. Because segment was able to disappear from currentlyMovingSegments, reduceLifetime logic was broken as well leading to potentially hidden stuck segments

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java [hunk 1]
```diff
@@ -91,9 +91,7 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
         params.getDruidCluster().getHistoricals().entrySet()) {
       String tier = entry.getKey();
 
-      if (currentlyMovingSegments.get(tier) == null) {
-        currentlyMovingSegments.put(tier, new ConcurrentHashMap<String, BalancerSegmentHolder>());
-      }
+      currentlyMovingSegments.putIfAbsent(tier, new ConcurrentHashMap<>());
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java]: Suggested computeIfAbsent()

**Comment by dgolitsyn** [comment]
No, this PR does not fix #4389, will create separate PR for it

**Comment by niketh** [comment]
@dgolitsyn @leventov This code will result in infinte wait for certain segements becuase of https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java#L98-L102. 

"all moved segments are removed by callbacks " - This is not true for cases like this one https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java#L169-L171

**Comment by dgolitsyn** [comment]
@niketh could you please explain how will segment in this case https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java#L169-L171 be added into `currentlyMovingSegments`?

**Comment by leventov** [comment]
@niketh I don't understand what do you mean.

**Comment by leventov** [comment]
@niketh any comments here?

**Comment by niketh** [comment]
@leventov By removing the else condition here https://github.com/druid-io/druid/pull/4472/files#diff-1e2aafb09822d416037bbcc45f163303L201, there are segments in the currentlyMovingSegments list which will never be removed. If they are not removed, they coordinator will never assign new segments because of this https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java#L99-L103 ?

Am I missing something?

**Comment by leventov** [comment]
@niketh segments are removed here: https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java#L185.

It's not that *all* segments *must* be removed in this loop: https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java#L122. Note that this loop is limited to `maxSegmentsToMove` steps, that comes from config and that generally could be less than `currentlyMovingSegments.size()`. It's supposed that `currentlyMovingSegments` becomes empty over several distinct calls to `DruidCoordinatorBalancer.run()`.

Also I'm not sure that it's actually bad if `currentlyMovingSegments` never becomes empty (because there are always more segments to move). It won't "reduce lifetimes" (I don't really understand what does it mean), but is that bad?

**Comment by niketh** [comment]
@leventov The callbacks that you pointed out are only registered for segments which match https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorBalancer.java#L170-L172 . 
For those that are in "else" have to be removed, and in the PR code that "else statement" has been removed

>> Also I'm not sure that it's actually bad if currentlyMovingSegments never becomes empty (because there are always more segments to move). It won't "reduce lifetimes" (I don't really understand what does it mean), but is that bad?
That is not bad, it just means that it will take more cycles to process the segments and new no assignments will be done till they are removed. Which is alright.

**Comment by leventov** [comment]
@niketh 
> For those that are in "else" have to be removed, and in the PR code that "else statement" has been removed

It's not a "side effect" of this PR, it's actually the very core purpose of this PR that is reflected in the title: **Do not remove segments from currentlyMovingSegments in DruidBalancer if move is impossible or not needed**. We think it's wrong to remove segments in this case. Why do you think segments "have to be" removed in this case?

**Comment by niketh** [comment]
@leventov 

> We think it's wrong to remove segments in this case. Why do you think segments "have to be" removed in this case?
Because if currentlyMovingSegments is not empty a new assignment cycle of DruidBalancer will not happen

**Comment by niketh** [comment]
If it's tested and it works fine, then 👍

**Comment by leventov** [comment]
@niketh this is tested, so merging

**Pull request merged**